clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -253,12 +257,229 @@ static void validatePackoffset(Sema &S, HLSLBufferDecl *BufDecl) {
+      if (requiresImplicitBufferLayoutStructure(Ty->getAsCXXRecordDecl()))
+        return true;
+    } else if (Ty->isConstantArrayType()) {
+      if (isZeroSizedArray(cast<ConstantArrayType>(Ty)))
+        return true;
+    }
+  }
+  // check bases
+  for (const CXXBaseSpecifier &Base : RD->bases())
+    if (requiresImplicitBufferLayoutStructure(
+            Base.getType()->getAsCXXRecordDecl()))
+      return true;
+  return false;
+}
+
+static CXXRecordDecl *findRecordDecl(Sema &S, IdentifierInfo *II,
+                                     DeclContext *DC) {
+  DeclarationNameInfo NameInfo =
+      DeclarationNameInfo(DeclarationName(II), SourceLocation());
+  LookupResult R(S, NameInfo, Sema::LookupOrdinaryName);
+  S.LookupName(R, S.getScopeForContext(DC));
+  if (R.isSingleResult())
+    return R.getAsSingle<CXXRecordDecl>();
+  return nullptr;
+}
+
+// Creates a name for buffer layout struct using the provide name base.
+// If the name must be unique (not previously defined), a suffix is added
+// until a unique name is found.
+static IdentifierInfo *getHostLayoutStructName(Sema &S,
+                                               IdentifierInfo *NameBaseII,
+                                               bool MustBeUnique,
+                                               DeclContext *DC) {
+  ASTContext &AST = S.getASTContext();
+  std::string NameBase;
+  if (NameBaseII) {
+    NameBase = NameBaseII->getName().str();
+  } else {
+    // anonymous struct
+    NameBase = "anon";
+    MustBeUnique = true;
+  }
+
+  std::string Name = "__layout_" + NameBase;
+  IdentifierInfo *II = &AST.Idents.get(Name, tok::TokenKind::identifier);
+  if (!MustBeUnique)
+    return II;
+
+  unsigned suffix = 0;
+  while (true) {
+    if (suffix != 0)
+      II = &AST.Idents.get((llvm::Twine(Name) + "_" + Twine(suffix)).str(),
+                           tok::TokenKind::identifier);
+    if (!findRecordDecl(S, II, DC))
+      return II;
+    // declaration with that name already exists - increment suffix and try
+    // again until unique name is found
+    suffix++;
+  };
+}
+
+// Returns true if the record type is an HLSL resource class
+static bool isResourceRecordType(const Type *Ty) {
+  return HLSLAttributedResourceType::findHandleTypeOnResource(Ty) != nullptr;
+}
+
+static CXXRecordDecl *createHostLayoutStruct(Sema &S, CXXRecordDecl *StructDecl,
+                                             HLSLBufferDecl *BufDecl);
+
+// Creates a field declaration of given name and type for HLSL buffer layout
+// struct. Returns nullptr if the type cannot be use in HLSL Buffer layout.
+static FieldDecl *createFieldForHostLayoutStruct(Sema &S, const Type *Ty,
+                                                 IdentifierInfo *II,
+                                                 CXXRecordDecl *LayoutStruct,
+                                                 HLSLBufferDecl *BufDecl) {
+  if (Ty->isRecordType()) {
+    if (isResourceRecordType(Ty))
```

Comment:
Won't we want to catch more intangible types than just resources here?  How do we make sure this is kept up-to-date for all intangible types?

On a related note, how do we ensure this code under `createHostLayoutStruct` always has matching/consistent logic to the code in `requiresImplicitBufferLayoutStructure`?  Could there be a way to share the key logic components perhaps?  For instance, there could be functions that return whether a leaf type is intangible (maybe this is just `Type::IsHLSLIntangibleType`), or is zero-sized, separate from the logic that recursively traverses a record type looking for these properties on field types.  The `IsHLSLIntangibleType` and zero-sized detection functions are then used by both traversal paths to ensure consistent behavior.

Reply:
Ok, I don't see cases like this tested:
```hlsl
namespace NS1 {
  struct Foo { ... };
  struct Bar { struct Foo { ... }; ... };
}
struct Foo { ... };
...
cbuffer CB1 {
  Foo foo1;
  NS1::Foo foo2;
  NS1::Bar::Foo foo3;
}

namespace NS2 {
  struct Foo { ... };
  cbuffer CB2 {
    ::Foo foo0;
    Foo foo1;
    NS1::Foo foo2;
    NS1::Bar::Foo foo3;
  }
}
```

In what decl context will implicit host layout decls be created for each type?  Will there be duplicates?

Reply:
There is an assert at the top of the `createHostLayoutStruct` function that will make sure the logic is consistent: 
```
assert(requiresImplicitBufferLayoutStructure(StructDecl) && "struct is already HLSL buffer compatible");
```

I have also moved the checking into a shared function `isInvalidConstantBufferLeafElementType`. Currently the only intangible types are resources classes, but I added a check for builtin resource type as well.